src/middlewared/middlewared/plugins/enterprise/m_series_nvdimm_firmware_freebsd.py

```diff
@@ -0,0 +1,69 @@
+# -*- coding=utf-8 -*-
+import logging
+import os
+import re
+
+import sysctl
+
+from middlewared.service import CallError, private, Service
+
+logger = logging.getLogger(__name__)
+
+
+class EnterpriseService(Service):
+
+    SIZE = None
+    SIZE_ERROR = "Size not retrieved yet"
+
+    VERSION = None
+    VERSION_ERROR = "Version not retrieved yet"
+
+    @private
+    def setup_m_series_nvdimm(self):
+        try:
+            self.SIZE = int(re.match("NVDIMM region ([0-9]+)GB", sysctl.filter("dev.nvdimm.0.%desc")[0].value).group(1))
+        except Exception as e:
+            self.SIZE_ERROR = str(e)
+
+        try:
+            # This should be ran before pool import so we cache the value obtained on first middlewared startup
+            cache_path = "/tmp/dev-nvdimm-0-dump_i2c"
+            if os.path.exists(cache_path):
+                with open(cache_path) as f:
+                    i2c_dump = f.read()
+            else:
+                i2c_dump = sysctl.filter("dev.nvdimm.0.dump_i2c")[0].value
+                with open(cache_path, "w") as f:
+                    f.write(i2c_dump)
+
+            prev_line = None
+            for line in i2c_dump.splitlines():
+                if prev_line == "Page 0x00:":
+                    version = line.split()[10]
+                    self.VERSION = f"{version[0]}.{version[1:]}"
+                    break
+                prev_line = line
+            else:
+                self.VERSION_ERROR = "Invalid dev.nvdimm.0.dump_i2c sysctl contents"
+        except Exception as e:
+            logger.error("Unable to retrieve M-Series NVDIMM firmware version", exc_info=True)
+            self.VERSION_ERROR = str(e)
+
+    @private
+    async def m_series_nvdimm_size(self):
+        if self.SIZE is None:
+            raise CallError(self.SIZE_ERROR)
+
+        return self.SIZE
+
+    @private
+    async def m_series_nvdimm_firmware_version(self):
+        if self.VERSION is None:
+            raise CallError(self.VERSION_ERROR)
+
+        return self.VERSION
+
+
+async def setup(middleware):
+    if (await middleware.call("truenas.get_chassis_hardware")).startswith("TRUENAS-M"):
```

Comment:
Same comment as above.
```suggestion
    if await middleware.call("failover.hardware") == "ECHOWARP":
```